src/main/application-update/check-for-updates/process-checking-for-updates.injectable.ts

```diff
@@ -0,0 +1,94 @@
+import updatesAreBeingDiscoveredInjectable from "../../../common/application-update/updates-are-being-discovered/updates-are-being-discovered.injectable";
+import discoveredUpdateVersionInjectable from "../../../common/application-update/discovered-update-version/discovered-update-version.injectable";
+import { runInAction } from "mobx";
+import assert from "assert";
+import askBooleanInjectable from "../../ask-boolean/ask-boolean.injectable";
+import quitAndInstallUpdateInjectable from "../../electron-app/features/quit-and-install-update.injectable";
+import downloadUpdateInjectable from "../download-update/download-update.injectable";
+import broadcastChangeInUpdatingStatusInjectable from "./broadcast-change-in-updating-status.injectable";
+import checkForUpdatesStartingFromChannelInjectable from "./check-for-updates-starting-from-channel.injectable";
+
+const processCheckingForUpdatesInjectable = getInjectable({
+  id: "process-checking-for-updates",
+
+  instantiate: (di) => {
+    const askBoolean = di.inject(askBooleanInjectable);
+    const quitAndInstallUpdate = di.inject(quitAndInstallUpdateInjectable);
+    const downloadUpdate = di.inject(downloadUpdateInjectable);
+    const selectedUpdateChannel = di.inject(selectedUpdateChannelInjectable);
+    const broadcastChangeInUpdatingStatus = di.inject(broadcastChangeInUpdatingStatusInjectable);
+    const checkingForUpdatesState = di.inject(updatesAreBeingDiscoveredInjectable);
+    const discoveredVersionState = di.inject(discoveredUpdateVersionInjectable);
+    const checkForUpdatesStartingFromChannel = di.inject(checkForUpdatesStartingFromChannelInjectable);
+
+    return async () => {
+      broadcastChangeInUpdatingStatus({ eventId: "checking-for-updates" });
+
+      runInAction(() => {
+        checkingForUpdatesState.set(true);
+      });
+
+      const { updateWasDiscovered, version, actualUpdateChannel } =
+        await checkForUpdatesStartingFromChannel(selectedUpdateChannel.value.get());
+
+      if (!updateWasDiscovered) {
+        broadcastChangeInUpdatingStatus({ eventId: "no-updates-available" });
+
+        runInAction(() => {
+          discoveredVersionState.set(null);
+          checkingForUpdatesState.set(false);
+        });
+
+        return;
+      }
+
+      broadcastChangeInUpdatingStatus({
+        eventId: "download-for-update-started",
+        version,
+      });
+
+      runInAction(() => {
+        // TODO: Unacceptable damage caused by strict mode
+        assert(version);
+        assert(actualUpdateChannel);
+
+        discoveredVersionState.set({
+          version,
+          updateChannel: actualUpdateChannel,
+        });
+
+        checkingForUpdatesState.set(false);
+      });
+
+      // Note: intentional orphan promise to make download happen in the background
+      downloadUpdate().then(async ({ downloadWasSuccessful }) => {
```

Comment:
`downloadUpdate` handles the possibility of reject inside for the code that requires it. So I'd say that it's not required.

Reply:
Then wouldn't `.catch(noop)` be a "readable code" mechanism for the same thing? Also, mixing `async` executors within `Promise` methods is a bad idea in general.

Reply:
> Then wouldn't .catch(noop) be a "readable code" mechanism for the same thing? 

Why would you want to add boilerplate when you know that it's not required? Code can be organized in a way that all for the catches in operational code are redundant boilerplate. If we start doing that they will became "new normal", so that you add them just because you want to be sure (you are not sure if it's required or not). This leads to more complex code than necessary. There should be unit test for all possible scenarios making all of these redundant. 

**Can you explain what is the scenario that you are worried about?** We can write unit test for that scenario to make sure that the application works as expected.

> Also, mixing async executors within Promise methods is a bad idea in general.

I do agree that `new Promise(async resolve => { ... })` is bad idea. But I'd like to know more about the fact that you are calling `somePromise.then(async value => { .... })`  a bad idea. What is the downside? Maybe some article that I can get familiar with? Generally `.then` can and should be avoided, but sometimes it's not the case.

Reply:
`somePromise.then(async value => { .... })` is mixing code styles unnecessarily, making the code harder to understand.

Reply:
I'd say usage of `iife` is more WTF than usage on `.then`.

Reply:
Resolved for the non-trivial parts of the conversation being addressed. I think.

Reply:
> I'd say usage of async iife is more WTF than usage of .then.

Why, it is a well documented code style? Even on MDN and has the benefits of `async` functions correctly handling all `await` points.

Plus you didn't address the concern of mixing code styles.

Reply:
> Why? It is a well documented code style; even on MDN.

Semantics, preference, really irrelevant here. It doesn't make any difference here one way or the other. As said before, function called here doesn't throw.

 > Plus you didn't address the concern of mixing code styles.
 
 Code style is irrelevant for the topic that we were discussing. You changed the topic mid conversation without addressing the real question that I had for you:

> Can you explain what is the scenario that you are worried about? We can write unit test for that scenario to make sure that the application works as expected.

Again, if you are worried about the unhandled promise rejection, could you answer the question I asked earlier?

Please, let's try focus on things that matter. I do see that this is not a good time investment.

Reply:
> No this is about code style as well, always has been.

Code style is not that important discussion to have in code reviews. There is much more important things that should be addressed in PR reviews. Sure, it's worth to mention, but not worth of going back and forth.

> As for what am I concerned about. I am concerned that a change to downloadUpdate could cause a floating promise to reject unexpectedly

Have you read the unit tests? _I know, it's big PR and these are hard to find but TBH tests should always be the most important part of the PR to be reviewed, since they describe behaviour and there might be something that needs to be iterated on that. This is another topic for the future, let's not have it here. :)_

1. `downloadPlatformUpdate.test.ts`  which tests the electron part of the implementation. Makes sure that Electron doesn't throw unexpectedly.
2. `installing-update.test.ts` which tests what happens when download fails.

Sure, there might be cases where our own implementation is done in a way that it will throw. But there is no way of being completely safe from that and it's just a good thing to crash completely. Amount of these cases is very minimal, when you have real unit tests that make sure you are not breaking something by doing something that you are not supposed to do. 

User input is more difficult topic, but the real answer is to think about all of the possible input types and add unit tests. E.g. we expect user to input string, what happens when user inputs number. Case for unit test, not for some arbitrary defensive code in all usages.

>  I don't see the problem with having another layer of defense if something else goes wrong.

Leads to place where developer just wraps everything in try-catch because she/he doesn't understand if there is code that throw. This pattern is currently present in the codebase and it's not good. 

Much better alternative is when usage of try-catch is limited to place where it's required, e.g. sending a HTTP request. Key thing here is that only the line sending the actual request should be try-catched. And the requirement for that comes from unit tests describing reason for it. Same goes for `.catch` obviously.

In that way, we don't need extra defensive code in usages. That makes our code less complex and reduces the cognitive load since you don't need to worry about possibility of thrown exceptions.

Reply:
I'm thinking about option C where we could decorate a function returning a promise as `withOrphanPromise`. That would: 
1. End conversation about code style
2. Be explicit in intent (i.e. no comment about intentional orphan promise required)
3. Give us a central place to add error handling to even the orphaned promises

It would look something like this:

```typescript
withOrphanPromise(downloadUpdate)();
```

Reply:
If that is the entire code then sure.